File: tests/rust/game/events.rs
//! The receive side: subscribe + drain, decoded payloads, and the
//! plugin-wide inbox fanning out to EVERY live script (the suite
//! registers two instances of this struct under different component
//! names — both must see both deliveries of one tick).

use super::util;
use crate::labelle::{self, EntityId, Script};

pub struct EventCounter {
    component: &'static str,
    e: EntityId,
    count: u32,
    amount: i64,
    nested_ok: bool,
}

impl EventCounter {
    pub fn new(component: &'static str) -> EventCounter {
        EventCounter {
            component,
            e: 0,
            count: 0,
            amount: 0,
            nested_ok: false,
        }
    }

    fn write(&self) {
        labelle::set_component(
            self.e,
            self.component,
            &format!(
                "{{\"amount\":{},\"count\":{},\"nested_ok\":{}}}",
                self.amount, self.count, self.nested_ok
            ),
        );
    }
}

impl Script for EventCounter {
    fn init(&mut self) {
        self.e = labelle::create_entity();
        labelle::subscribe("cargo__delivered");
        self.write();
    }

    fn on_event(&mut self, name: &str, payload: &str) {
        if name != "cargo__delivered" {
            return;
        }
        self.count += 1;
        self.amount = util::i64_field(payload.as_bytes(), "\"amount\":").unwrap_or(-1);
        // Nested payloads arrive intact (a structural spot-check — full
        // decoding is the script's own business in slice 1).
        self.nested_ok = payload.contains("\"tags\":[\"fragile\"]");
        self.write();
    }
}

File: tests/go/game/events.go
// The inbox fan-out double: two instances of this script subscribe to
// the same event; every drained entry must reach BOTH (the inbox is
// plugin-wide), with the nested payload intact.
package game

import (
	"fmt"

	"labelle"
)

type EventCounter struct {
	labelle.NoopScript
	comp   string
	entity labelle.EntityID
	count  int
	amount int64
	nested bool
}

func NewEventCounter(comp string) *EventCounter {
	return &EventCounter{comp: comp}
}

func (e *EventCounter) write() {
	labelle.SetComponent(e.entity, e.comp,
		fmt.Sprintf(`{"amount":%d,"count":%d,"nested_ok":%v}`, e.amount, e.count, e.nested))
}

func (e *EventCounter) Init() {
	e.entity = labelle.CreateEntity()
	labelle.Subscribe("cargo__delivered")
	e.write()
}

func (e *EventCounter) OnEvent(name, payload string) {
	if name != "cargo__delivered" {
		return
	}
	e.count++
	if v, ok := i64Field([]byte(payload), `"amount":`); ok {
		e.amount = v
	}
	e.nested = contains(payload, `"tags":["fragile"]`)
	e.write()
}
